syntax = "proto3";

package WW;

option cc_generic_services = true;

// 日志条目
message LogEntry {
    int32 index = 1;                // 日志索引
    int32 term = 2;                 // 日志任期
    bytes command = 3;              // 日志内容
}

// 选举投票请求
message RequestVoteRequest {
    int32 term = 1;                 // 候选人任期
    int32 candidate_id = 2;         // 候选人 ID
    int32 last_log_index = 3;       // 候选人最后日志条目索引
    int32 last_log_term = 4;        // 候选人最后日志条目任期
}

// 选举投票响应
message RequestVoteResponse {
    int32 term = 1;                 // 当前节点任期
    bool voted = 2;                 // 当前节点是否投票
}

// 日志条目请求
message AppendEntriesRequest {
    int32 term = 1;                 // Leader 任期
    int32 leader_id = 2;            // Leader ID
    int32 prev_log_index = 3;       // 上一个日志索引
    int32 prev_log_term = 4;        // 上一个日志任期
    repeated LogEntry entry = 5;    // 日志条目，为空则是心跳包
    int32 leader_commit = 6;        // Leader 已提交索引
}

// 日志条目响应
message AppendEntriesResponse {
    int32 term = 1;                 // 当前节点任期
    bool success = 2;               // 是否成功追加
}

// Raft 服务
service RaftService {
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
}
